### PR TITLE
Remove Path.parentPaths()

### PR DESCRIPTION
**Pull request by kendrickt**
The order of paths was very specific and felt dangerous to use, and it was only used in one place (ApplicantData.put) which can be done with recursion.

### Description
Path.parentPaths() is removed in favor for recursion, which is not dependent on the order of paths. 

### Checklist
- [ ] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

### CODE REVIEW

File: universal-application-tool-0.0.1/app/services/applicant/ApplicantData.java [hunk 1]
```diff
@@ -120,16 +120,10 @@ private void putNull(Path path) {
...
-        .forEach(
-            segmentPath -> {
-              if (!hasPath(segmentPath)) {
-                this.jsonData.put(
-                    segmentPath.parentPath().path(), segmentPath.keyName(), new HashMap<>());
-              }
-            });
-
-    this.jsonData.put(path.parentPath().path(), path.keyName(), value);
+    if (!hasPath(path.parentPath())) {
```
Comment (carolinedanzi) [inline universal-application-tool-0.0.1/app/services/applicant/ApplicantData.java:123]: Neat!

I know "applicant" is pretty much in every path, but I think we should still include a base case here. If for some reason the path doesn't start with "applicant", eventually you'll get to Path.empty(), which will throw an error like "Json cannot be null or empty"

Comment (kendrickt) [reply universal-application-tool-0.0.1/app/services/applicant/ApplicantData.java:123]: Ahh thanks! I think that may be why all the other tests are failing.

Comment (kendrickt) [reply universal-application-tool-0.0.1/app/services/applicant/ApplicantData.java:123]: It was not! :(

File: universal-application-tool-0.0.1/app/services/Path.java [hunk 2]
```diff
@@ -42,12 +42,22 @@ private static Path create(ImmutableList<String> pathSegments) {
    */
   public abstract ImmutableList<String> segments();
 
+  public boolean isEmpty() {
```
Comment (carolinedanzi) [inline universal-application-tool-0.0.1/app/services/Path.java:45]: super nit: may as well `@Memoize` this too

File: universal-application-tool-0.0.1/app/services/applicant/ApplicantData.java [hunk 3]
```diff
@@ -120,7 +120,7 @@ private void putNull(Path path) {
    * @param value the value to place; values of type Map will create the equivalent JSON structure
    */
   private void put(Path path, Object value) {
-    if (!hasPath(path.parentPath())) {
+    if (!path.parentPath().isEmpty() && !hasPath(path.parentPath())) {
```
Comment (carolinedanzi) [inline universal-application-tool-0.0.1/app/services/applicant/ApplicantData.java:123]: nice! suggestion (blocking): mind adding a quick test for this case, just to be safe?

**Review by carolinedanzi — APPROVED** [review]

**Pull request merged**